### ISSUE TITLE
dual / multiple fuel maps

### ISSUE THREAD

**Issue by rusefillc** [rusefi/rusefi#3114]
popular question is dual / multiple fuel maps

Владимир 5634 from whatsapp elaborates that this is about multiple boost modes in his specific case

**Comment by rusefillc** [comment]
Depends on #3045

**Comment by mck1117** [comment]
obligatory X-Y problem question: what are users expecting to change at runtime that will actually change the fueling such that it needs a full separate VE table?

**Comment by rusefillc** [comment]
Pavel says he needs two maps:
1) eco mode, up to 4k highway driving
2) sport mode

NA engines

**Comment by mck1117** [comment]
But why would the *VE MAP* be different? What does he actually want to change? There should be no change to the VE map based on what the driver wants.

His desire sounds like he might actually want to *ETB pedal maps*, not two *VE maps*

**Comment by shadowm60** [comment]
In eco you need higher afr, you want less fuel consumtion, in sport you want different target afr. I thing there should be two target afr tables, two ve and two ignition maps. Based on a switch select the mode?

**Comment by mck1117** [comment]
I disagree.  Your target AFR table should be shaped so that you're in an efficient regime while cruising, and a best-torque regime while floored.  Ignition is even easier: everywhere should use the same ignition timing.  Ideal power and ideal efficiency are *identical* for ignition.  There certainly shouldn't be two VE tables - as that's a fundamental constant of the engine, not anything to do with how you want the car to drive.

I doubt any OEM actually changes anything other than pedal->tps relationship and boost target in different modes. Target AFR, timing, and VE have no reason to change.

**Comment by AlexHLinS** [comment]
The different VE map needed when VE dramatically changed by smth, for example vtec on/off, Vvti etc. Two maps sport and eco will realize only by accel pedal reaction or target boost (but we talking about NA). Also changing ignition map, afr map etc is not an solution, coz best mpg <-> max torq

**Comment by dron0gus** [comment]
We need separate limp mode VE table (TPS/RPM).
I think it will not look exactly the same as MAP VE table?

We (at least in Russia) need few AFR/advance tables due to absence of good/sport fuel on each gas station.

**Comment by rusefillc** [comment]
Simon - Orchardperformance  7:25 PM
I agree with your comments. Correct tuning for mbt is correct for mbt and stay out of enrichment ranges.
Its something that comes up all the time.
Doing something to suit everyone...
Back to the main map + modification table idea, if we make the reason for switching to the modification table user configurable then people can set it to what ever they want.
The become just condition A for main table, condition B for 100% modifiction table.
Doesn't matter if condition B is eco mod, NOS, high boost, race gas or mother in law mode.





7:26
Honestly I think just adding maybe 4 selectable 8x8 modification tables and using them for VVT, eth, map switching, etc is the best catch all that kills a lot of targets in one go
7:28
On the other side, cam timing changes in eco mode but its not a VE map change its a total engine operating point change. Like for like the VE at that operating condition is still the same regardless of mode.

Matthew Kennedy  7:43 PM
There might be technical challenges with that owing to the fact that we use tunerstudio
7:43
Link ecus have some memory region for table storage, and you can configure which function uses which tables
7:44
But the software knows about that
7:44
There's no way, afaik, to tell TS to switch WHERE a table is located based on another setting

Simon - Orchardperformance  8:07 PM
Hmmm
8:11
How does FSIO handle that curently?
Is there something that acts as a place holder within the various calculations, so instead of having to point to a table we point the table to the place holder?

Dave B.  1:47 AM
Tunerstudio... 90% awesome

Matthew Kennedy  1:48 AM
@Simon - Orchardperformance Fsio doesn't handle that

Simon - Orchardperformance  8:11 AM
Just trying to understand the nature of the issue, but does tuner studio need to know about it at all?

Simon - Orchardperformance  8:18 AM
I guess my point is, do the tables need to switch where it is located?
If we take a situation where we were going to add 2x 16x16 Timing and Fuel tables like an old dual table ECU, if those tables were broken into 8x 8x8 tables then we could allocate 2 as a AFR modifier, 2 as a VE modifier and 4 as a timing modifier for the same amount of RAM?
If we put some "hooks" into the calculations we would want to modify with them and set the default action of those hooks to be "no modification" then surely all we would need TS to do is to configure if that hook is doing nothing or bringing in the modification.
Think a drop down with "no modification", "+1 dimension", "+2 dimensions".
8:20
If the issue is about being able to choose the trigger of the modification then I would have thought existing systems would be able to do that, If each table has a 2d "bias" table, like the one we were saying is a % to % style mapping, then is it not simply a case of switching the axis ID for that table?

Simon - Orchardperformance  8:26 AM
Say if we want to blend from the main timing table to "timing modifier 1" based on flex fuel sensor then we set in TS the setting for "+1 dimension" which activates the calculation that adds the value of the modification table * bias table to the spark demand.
We then set "bias table 1" axis to be flex fuel and that allows 0->100% of the additional "modification table 1"
8:28
I know that idea limits us to only a couple of VE, AFR and Spark modification but even then, wielded well I cannot see how someone would want to try tuning more than 6 dimensions of fuel modification, it would be a mess.
8:30
I guess my TLDR there is "do we need to switch anything, just simply activate or ignore its affect on the calculation"

Simon - Orchardperformance  8:38 AM
I am also happy to accept that I am completely misunderstanding the problem statement

BeerMoneyMotorsports  11:29 AM
image.png 
image.png


11:29
do we really need both?

Simon - Orchardperformance  2:03 PM
Probably yes
2:04
Although they do similar things one is MUCH simpler than the other

Matthew Kennedy  2:04 PM
we need something TPS based

Simon - Orchardperformance  2:04 PM
in terms of tuning and function
2:05
What we could probably do with is some dedicated testing time spent on the TPS method and some documentation writing. I am not 100% the TPS/TPS is working properly.

Matthew Kennedy  2:05 PM
I don't think it does.
2:05
I'd love to build something that has a smaller parameter space
2:06
because that table is fucking hard to tune even if it works

Simon - Orchardperformance  2:06 PM
yeahhhh

Simon - Orchardperformance  2:15 PM
at a minimum we need a way to compensate that initial hit before the wall wetting can take over and that means TPS based.

Matthew Kennedy  2:15 PM
yes, that's what I mean

Simon - Orchardperformance  2:15 PM
I suspect that table could be reduced in size quite a bit and still be effective. iirc old motronic is only a 4x4

Matthew Kennedy  2:16 PM
when you crack the throttle air is gonna find its way to the port awful fast, before the map sensor can get there

Simon - Orchardperformance  2:16 PM
yep
2:17
I wonder if there is a clever way to hook into the wall wetting and introduce a proportional amount based on the WW settings when a TPS change is seen

Matthew Kennedy  2:18 PM
hmm

Simon - Orchardperformance  2:18 PM
i.e. sees TPS change, introduces a proportional amount, then deactivates as WW comes in

Matthew Kennedy  2:18 PM
one idea I had was to use a "lookahead MAP" with the expected MAP table
2:19
est_map = map + high_pass(alphan_map)

Simon - Orchardperformance  2:20 PM
throwing this out/thinking out loud.... The % area increase curve is a fixed thing for a round throttle regardless of the diameter

Matthew Kennedy  2:20 PM
yeees
2:21
you're heading in the direction I like

Simon - Orchardperformance  2:21 PM
we could estimate a % increase of air flow from the TPS change based on the current air mass
2:21
with just a 1x8 table to do a bit of correction
2:22
for say if you have a non round throttle or a double barrel that staggers the throttle plates
2:23
perhaps we can try pre-empting the wall wetting MAP change using that?
2:23
afterall it does not have to be perfect, it probably only functions for at most 2 combustions on transition

Matthew Kennedy  2:23 PM
we could even just put that in config, no correction table, defaulting to "round throttle with modestly sized shaft"

Simon - Orchardperformance  2:24 PM
shaft won't matter as its effect is needed at low angles

Matthew Kennedy  2:24 PM
fair

BeerMoneyMotorsports  2:24 PM
shaft always matter

Simon - Orchardperformance  2:24 PM
says the man talking to "dutch miata groups" (edited) 

BeerMoneyMotorsports  2:25 PM
is there something i should know about dutch miata clubs?

Matthew Kennedy  2:25 PM
well they're dutch
2:25
and they can't /all/ be hairdressers

Simon - Orchardperformance  2:25 PM
Is it like "Mongolian basket weaving forums"?

BeerMoneyMotorsports  2:26 PM
please go back to brainstorming and making @Matthew Kennedy excited

Simon - Orchardperformance  2:26 PM
https://www.reddit.com/r/OutOfTheLoop/comments/4308d6/why_are_people_referring_to_4chan_as_a/

redditreddit
Why are people referring to 4chan as a "Bangladeshi Kite testing forum", "Chinese frog breeding forum", and so on?
This seems to be a recent thing. [Example](https://www.reddit.com/r/4chan/comments/41k0qu/jeb_makes_a_thread/cz3aju4?context=10)
2:27
Just in case that made no sense

BeerMoneyMotorsports  2:27 PM
thank you, it surely did not make sense.  lol
2:27
now i get it

Simon - Orchardperformance  2:27 PM
a "russian ecu building forum"

BeerMoneyMotorsports  2:28 PM
the reference name for this group with my friends is russian spyware

Matthew Kennedy  2:28 PM
for mine it's "that communist ECU thing"

BeerMoneyMotorsports  2:29 PM
our friends comrade

Simon - Orchardperformance  2:31 PM
a slavic alkene injection support group

Matthew Kennedy  2:33 PM
ah yes
2:33
anyway
2:33
throttley goodness

Simon - Orchardperformance  2:33 PM
I do kind of wonder if "dutch miata" is a bit like a dutch oven but instead you fart in your mates miata, get out and hold the door shut?
anyway....

BeerMoneyMotorsports  2:34 PM
-jots down notes for dutch miata

Simon - Orchardperformance  2:35 PM
https://www.youtube.com/watch?v=R1DwrLTYeeI

YouTubeYouTube | VeryCoolCat
Guy farts in a singing booth 

2:35
We probably need some kind of correction table in there simply because if we don't someone will hit an edge case where it causes hesitation
2:38
I think the big challenge will be estimating the size of the effect that RPM has on the throttle change.
2% change at low RPM is more significant MAP change than at higher

Matthew Kennedy  2:38 PM
but if we know how much airflow, and in which "part" of the throttle's travel, then we can estimate that

Simon - Orchardperformance  2:38 PM
but I guess if we know how much air we are currently consuming then we can make an estimate based on RPM of the size of that effect
2:40
yep, i.e. if we are say 50% VE at low rpm and we increase from 5-10% airflow area we know we are likely to increase to 100% VE
2:41
If we litterally ignore the whole angle thing and just consider area available to flow air through we can probably get around that
2:41
end of the day I know we can abstract the throttle down to a simple orifice and get reasonable results, we do it all the time in engine simulation work
2:42
and all that matters is the size of that orifice and the coefficient of discharge, which will change with angle so perhaps that is where the correction table would help some
2:44
the low degree opening angles are much more sensitive simply because 1 degree angle can be over 100% area increase
2:45
Infact we an check this theory out with a couple of math channels in a data log

Matthew Kennedy  3:13 PM
@Simon - Orchardperformance @Dave B. I did some thinking while backpacking about table switching. I think I have a viable solution thought up.
Instead of "VE table 1" and "VE table 2", just have "VE table" (used by default), then N "switchable tables" that can get mapped in place of some other table by lua
3:14
Something like
switchVeTable(2) switches to use switchable 2 in place of ve, and switchVeTable(0) uses the default

Zach Hoobler  3:19 PM
I think that's basically what link does?
3:19
from what I've seen of link it appears they just have a bunch of tables and then you configure which tables to use for what

Matthew Kennedy  3:20 PM
This is sort of that but gets around the issue of TS being unable to handle that.

Zach Hoobler  3:21 PM
Yeah this has a default assigned table and then you have override tables
3:21
so the base table is fixed, but it's fine because 99.99% of people will need that VE table

Simon - Orchardperformance  4:47 PM
That would probably work, I suspect that also makes it easy to make the tables modifier tables.

Zach Hoobler  4:49 PM
presumably a user could choose to use those as modifiers OR complete replacements

Dave B.  5:19 PM
this sounds appealing

Jim Rowan  7:44 PM
seems to me that 'modifier' is way more complicated.  'switch' means change the address.  'modifier' means add an extra address and change the algorithm.  of course, it's just software.  :slightly_smiling_face:

Zach Hoobler  7:46 PM
I see modifier as "add these tables together when modifier enabled"
7:47
not really all that different from "replace this table with that one"

Matthew Kennedy  7:47 PM
then you have to support blending
7:47
base + (x1 * table1) + (x2 * table2) + ... + (xn * tablen)
7:47
which is ok
7:50
@Dave B. that give you any flashbacks to ford mapped point vvt shit?
:eyes:
1


Simon - Orchardperformance  8:00 PM
Supporting blending will be needed anyway
8:00
At least for handling the cam transitions and multi fuel

Matthew Kennedy  8:00 PM
for flex at least, yes

**Comment by rusefillc** [comment]
@NMSTEC is asking for dual VE and timing table since front and real cylinders on HD behave differently due to V-twin and cooling and whoknows what else

**Comment by AlexHLinS** [comment]
Could we fix it by individual cylinder trim?

**Comment by mck1117** [comment]
We're conflating two different problems/features here.

1. Switchable/blendable maps
2. Per-cylinder trim maps

**Comment by BeerMoneyMotorsports** [comment]
Can we close this and create a ticket that adds timing, cranking and etc tables?  Flex is a feature that’s becoming more common.  Fuel blending the ve table is fine but there are cranking and ignition table(that i know) are different between fuels.

**Comment by NMSTEC** [comment]
This is now a bit of a mess. This can also be taken as Dual Maps as I need them for enrichment. I'd say "User selectable performance maps" would be a better new ticket.

**Comment by rusefillc** [comment]
@NMSTEC please explain WHY and WHAT performance means separate maps?

**Comment by NMSTEC** [comment]
Multiple uses for this, and reasons.

1. Limp mode (Harley will need this, if meth runs out, overboost, excessive knock, CLT too high)
2. Different fuel map (Instead of running 91 & meth, when no available, select low octane map with retarded timing, and more enrichment)
3. Scramble mode (When you want much more power than normal, for a short period of time)
4. Different fuel map again, but this time for E85. It has massively different everything in regards to timing and fuel.

**Comment by mck1117** [comment]
How many of those cannot be done in lua?

1. Limp mode: subtract 0.2 from target lambda, subtract 5-10 degrees from timing
2. See 1
3. See 1, but do the opposite
4. E85 support is a somewhat separate problem, we already do automatic mixture adjustment for E85, and your target lambda will at least be close to the same as on gas. E85 timing blend is #3778 
5. Not a good reason!

**Comment by NMSTEC** [comment]
1. How would you do this in LUA without as you say, eating too much cpu time? Especially on the budget F4 HD is going to be running.
2. Fair enough.
3. This is stretching it.
4. -
5. Is good reason. What's a good reason for not doing this?

**Comment by duhafnusa4** [comment]
this needs updates

**Comment by rusefillc** [comment]
Josh Stewart says

We had a similar conversation with Speeduino. The most common arguments were around things like being able to switch to a different fuel (Eg LPG) without needing to manually change the tune and being able to do blended modes for setups such as boosted ITBs.
The dual tables with blending works really well for the latter now, so it ended up being a nice addition.

**Comment by biohazardav** [comment]
I don't understand why it's such a big argument to add switchable or blended fuel and spark tables. Many other engine management solutions offer this and is required for things like variable valve timing, VTEC and cammed/boosted engines with other switchable power adders. What would it take to make this happen?

**Comment by rusefillc** [comment]
hello from 2025!

blends are the way.

**Comment by rusefillc** [comment]
10 years in the making https://github.com/rusefi/rusefi/issues/9286
